packages/unified-selection/src/unified-selection/HiliteSetProvider.ts

```diff
@@ -0,0 +1,341 @@
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+
+/** @packageDocumentation
+ * @module UnifiedSelection
+ */
+
+import { ECClass, IMetadataProvider, parseFullClassName } from "./queries/ECMetadata";
+import { ECSqlBinding, IECSqlQueryExecutor } from "./queries/ECSqlCore";
+import { SelectableInstanceKey, Selectables } from "./Selectable";
+
+/**
+ * A set of model, subcategory and element ids that can be used for specifying hilite.
+ * @public
+ */
+export interface HiliteSet {
+  models: string[];
+  subCategories: string[];
+  elements: string[];
+}
+
+/**
+ * Props for creating a `HiliteSetProvider` instance.
+ * @public
+ */
+export interface HiliteSetProviderProps {
+  queryExecutor: IECSqlQueryExecutor;
+  metadataProvider: IMetadataProvider;
+}
+
+/**
+ * ECSQL based provider which determines what `HiliteSet`
+ * should be hilited based on the supplied `Selectables`.
+ * @public
+ */
+export class HiliteSetProvider {
+  private _queryExecutor: IECSqlQueryExecutor;
+  private _metadataProvider: IMetadataProvider;
+  // Map between a class name and its type
+  private _classRelationCache: Map<string, InstanceIdType>;
+
+  private constructor(props: HiliteSetProviderProps) {
+    this._queryExecutor = props.queryExecutor;
+    this._metadataProvider = props.metadataProvider;
+    this._classRelationCache = new Map<string, InstanceIdType>();
+  }
+
+  /**
+   * Create a hilite set provider.
+   */
+  public static create(props: HiliteSetProviderProps) {
+    return new HiliteSetProvider(props);
+  }
+
+  /**
+   * Get hilite set for supplied `Selectables`.
+   * @param selection Selection to get hilite set for.
+   */
+  public async getHiliteSet(selection: Selectables): Promise<HiliteSet> {
+    const models: string[] = [];
+    const subCategories: string[] = [];
+    const elements: string[] = [];
+
+    const iterator = this.getHiliteSetIterator(selection);
+    for await (const set of iterator) {
+      models.push(...set.models);
+      subCategories.push(...set.subCategories);
+      elements.push(...set.elements);
+    }
+
+    return {
+      models,
+      subCategories,
+      elements,
+    };
+  }
+
+  /**
+   * Get hilite set iterator for supplied `Selectables`.
+   */
+  public async *getHiliteSetIterator(selection: Selectables): AsyncIterableIterator<HiliteSet> {
+    // Map between element ID types and EC instance IDs
+    const keysByType = new Map<InstanceIdType, string[]>();
+
+    for (const entry of selection.instanceKeys) {
+      for (const key of entry[1]) {
+        await this.addKeyByType({ className: entry[0], id: key }, keysByType);
+      }
+    }
+
+    for (const entry of selection.custom) {
+      for await (const key of entry[1].loadInstanceKeys()) {
+        await this.addKeyByType(key, keysByType);
+      }
+    }
+
+    yield {
```

Comment:
This does not exactly work as iterator as in previous implementation it loads all ids and then emits them in one go.

Reply:
I think this would be an ideal solution:
- `getHilitedModels`, `getHilitedSubCategories` and `getHilitedElements` return `Observable<Id64String>`.
- we subscribe to all of them and accumulate the values for fixed time intervals - on each interval we emit an accumulated hilite set if it's not empty.

What do you think?

Reply:
I agree. We could also look into streaming instance keys from selectables instead of loading them all before running any query as `CustomSelectable.loadInstanceKeys` might produce a lot of keys.

So we can leave this as is for now and comeback to it later.

Reply:
Added a backlog item: https://github.com/iTwin/presentation/issues/501.